### PR TITLE
introduce `balanceCheck` and `balanceNotify`.

### PR DESCRIPTION
**Pull request by fiatjaf**
As discussed here: https://t.me/lnurl/9729

### CODE REVIEW

File: lnurl-withdraw.md [hunk 1]
```diff
@@ -13,12 +13,13 @@ Today users are asked to provide a withdrawal Lightning invoice to a service, th
...
     {
+        tag: "withdrawRequest", // type of LNURL
         callback: String, // the URL which LN SERVICE would accept a withdrawal Lightning invoice as query parameter
         k1: String, // random or non-random string to identify the user's LN WALLET when using the callback URL
-        maxWithdrawable: MilliSatoshi, // max withdrawable amount for a given user on LN SERVICE
         defaultDescription: String, // A default withdrawal invoice description
-        minWithdrawable: MilliSatoshi // An optional field, defaults to 1 MilliSatoshi if not present, can not be less than 1 or more than `maxWithdrawable`
-        tag: "withdrawRequest" // type of LNURL
+        maxWithdrawable: MilliSatoshi, // max withdrawable amount for a given user on LN SERVICE
+        minWithdrawable: MilliSatoshi, // An optional field, defaults to 1 MilliSatoshi if not present, can not be less than 1 or more than `maxWithdrawable`
```
Comment (Kixunil) [inline lnurl-withdraw.md:21]: Should be allowed to be zero. The use case is in ATMs which want to make sure users wallet supports this before allowing to insert money. This avoids having to process returns of money.

File: lnurl-withdraw.md [hunk 2]
```diff
@@ -36,13 +37,39 @@ Today users are asked to provide a withdrawal Lightning invoice to a service, th
...
 6. `LN SERVICE` sends a `{"status": "OK"}` or `{"status": "ERROR", "reason": "error details..."}` JSON response and then attempts to pay the invoices asynchronously.
 7. `LN WALLET` awaits for incoming payment if response was successful.
 
 Note that service will withdraw funds to anyone who can provide a valid ephemeral `k1`. In order to harden this a service may require autorization (LNURL-auth, email link etc.) before displaying a withdraw QR.
 
+### `balanceCheck`
+
+This _optional_ property is meant to enable automatic delivery of funds from services to wallets. It works as follows:
+
+  * If `LN WALLET` supports the `balanceCheck` scheme and sees a `balanceCheck` property from a withdraw it performs, it may store that URL locally and call it later (for example, at its next startup, or after 24 hours).
```
Comment (Kixunil) [inline lnurl-withdraw.md:55]: To complement the above, I'd add: "If the withdrawable amount is zero and the wallet does **not** support `balanceNotify`, then the wallet SHOULD poll the service once per minute for the following 30 minutes."

The idea is that zero means it's expected to go above zero ~soon. But maybe add a special flag for that?

**Review by Kixunil — COMMENTED** [review]
While this doesn't resolve #14 completely, it seems to solve the most important parts. Channel opening could be added later if desired.

File: lnurl-withdraw.md (hunk 1 above)
Comment (fiatjaf) [reply lnurl-withdraw.md]: I'm not sure about that very specific use case, but I think it makes sense to allow zero, specially considering automatic `balanceCheck`s.

File: lnurl-withdraw.md [hunk 3]
```diff
@@ -77,8 +77,8 @@ Eg:
 	https://LNserviceURL
 	?tag=withdrawRequest
 	&k1=String
-	&minWithdrawable=MilliSatoshi
-	&maxWithdrawable=MilliSatoshi
+	&minWithdrawable=Integer
+	&maxWithdrawable=Integer
```
Comment (Kixunil) [inline lnurl-withdraw.md:81]: I personally find specifying units much easier to understand. What is the reason for this change?

File: lnurl-withdraw.md [hunk 4]
```diff
@@ -17,8 +17,8 @@ Today users are asked to provide a withdrawal Lightning invoice to a service, th
         callback: String, // the URL which LN SERVICE would accept a withdrawal Lightning invoice as query parameter
         k1: String, // random or non-random string to identify the user's LN WALLET when using the callback URL
         defaultDescription: String, // A default withdrawal invoice description
-        maxWithdrawable: MilliSatoshi, // max withdrawable amount for a given user on LN SERVICE
-        minWithdrawable: MilliSatoshi, // An optional field, defaults to 1 MilliSatoshi if not present, can not be less than 1 or more than `maxWithdrawable`
+        minWithdrawable: Integer, // Min amount (in millisatoshis) the user can withdraw from LN SERVICE, or 0
```
Comment (Kixunil) [inline lnurl-withdraw.md:20]: Nit: "Min amount or 0" sounds a bit funny to me, I'd word it as "Min amount, can be 0"

File: lnurl-withdraw.md (hunk 3 above)
Comment (chill117) [reply lnurl-withdraw.md:81]: I agree. Especially since the lnurl-channel spec uses whole sats instead of msats.

Comment (fiatjaf) [reply lnurl-withdraw.md:81]: I just find it clearer this way because "MilliSatoshi" is not a valid JSON type. But I can change this back.

**Comment by fiatjaf** [comment]
Maybe a better and quicker explanation here: https://twitter.com/fiatjaf/status/1388216087817424899
